packages/server/modules/workspaces/repositories/workspaces.ts

```diff
@@ -35,26 +38,92 @@ export const upsertWorkspaceFactory =
       .merge(['description', 'logoUrl', 'name', 'updatedAt'])
   }
 
-export const getWorkspaceRoleFactory =
-  ({ db }: { db: Knex }): GetWorkspaceRole =>
+export const getWorkspaceRolesFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoles =>
+  async ({ workspaceId }) => {
+    return await tables.workspacesAcl(db).select('*').where({ workspaceId })
+  }
+
+export const getWorkspaceRoleForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRoleForUser =>
   async ({ userId, workspaceId }) => {
-    const acl = await tables
-      .workspacesAcl(db)
-      .select('*')
-      .where({ userId, workspaceId })
-      .first()
+    return (
+      (await tables
+        .workspacesAcl(db)
+        .select('*')
+        .where({ userId, workspaceId })
+        .first()) ?? null
+    )
+  }
 
-    return acl || null
+export const getWorkspaceRolesForUserFactory =
+  ({ db }: { db: Knex }): GetWorkspaceRolesForUser =>
+  async ({ userId }, options) => {
+    const workspaceIdFilter = options?.workspaceIdFilter ?? []
+
+    const query = tables.workspacesAcl(db).select('*').where({ userId })
+
+    if (workspaceIdFilter.length > 0) {
+      query.whereIn('workspaceId', workspaceIdFilter)
+    }
+
+    return await query
+  }
+
+const isUserLastWorkspaceAdmin = (
+  workspaceRoles: WorkspaceAcl[],
+  userId: string
+): boolean => {
+  const workspaceAdmins = workspaceRoles.filter(
+    ({ role }) => role === 'workspace:admin'
+  )
+  const isUserAdmin = workspaceAdmins.some((role) => role.userId === userId)
+
+  return isUserAdmin && workspaceAdmins.length === 1
+}
+
+export const deleteWorkspaceRoleFactory =
+  ({ db }: { db: Knex }): DeleteWorkspaceRole =>
+  async ({ userId, workspaceId }) => {
+    // Protect against removing last admin in workspace
+    const workspaceRoles = await getWorkspaceRolesFactory({ db })({ workspaceId })
+
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
+      throw new Error('Cannot remove last admin in workspace.')
+    }
+
+    // Bail early if user has no role
+    const currentRole = workspaceRoles.find((role) => role.userId === userId)
```

Comment:
instead of running a sql query against the db, inspecting the return value and based on a condition maybe run a second, you can issue a delete command with a returning '*' statement, to get back the deleted rows.

This way you are guaranteed to only run one sql statement.

https://knexjs.org/guide/query-builder.html#returning